### PR TITLE
Use ray dataset and drop type casting in binary_feature prediction post processing for speedup

### PR DESCRIPTION
**Pull request by magdyksaleh**
# Code Pull Requests

Update the binary feature post-processing to use ray datasets to speed up overall time 

# Documentation Pull Requests

Note that the documentation HTML files are in `docs/` while the Markdown sources are in `mkdocs/docs`.

If you are proposing a modification to the documentation you should change only the Markdown files.

`api.md` is automatically generated from the docstrings in the code, so if you want to change something in that file, first modify `ludwig/api.py` docstring, then run `mkdocs/code_docs_autogen.py`, which will create `mkdocs/docs/api.md` .

### CODE REVIEW

**Review by tgaddair — COMMENTED** [review]
@magdyksaleh I'm going to suggest a different approach here, since I suspect the issue we're seeing here with `binary_feature` will be an issue for other output feature types as well.

The key idea is that all of the `postprocess_predictions` functions for all output features are only applying per-row transformations. There are not transformations being applied that require computing column-level metadata in this step. As such, instead of passing Dask DFs to the postprocess functions, we can run one top-level `map_batches` call that applies the transformations to batches as pandas dataframes.

Concretely:

1. Change the `postprocess` function in `postprocessing.py` to something like this:

```
def posproc_preds(df):
        for of_name, output_feature in output_features.items():
            return output_feature.postprocess_predictions(
                df,
                training_set_metadata[of_name],
            )
    
    predictions = backend.df_engine.map_batches(predictions, posproc_preds)
```

2. Change every `postprocess_predictions` function by removing `output_directory` and `backend` from the signature.

3. Change all the calls to `backend.df_engine.map_objects` in `postprocess_predictions` to simply call `df.map`, since it's just a Pandas DF.

File: ludwig/data/dataframe/base.py
Comment (tgaddair) [inline ludwig/data/dataframe/base.py:50]: I would rename to just `map_batches`, and maybe drop the `batch_format` param, since it would be difficult to make it work with other backends.

**Comment by magdyksaleh** [comment]
> @magdyksaleh I'm going to suggest a different approach here, since I suspect the issue we're seeing here with `binary_feature` will be an issue for other output feature types as well.
> 
> The key idea is that all of the `postprocess_predictions` functions for all output features are only applying per-row transformations. There are not transformations being applied that require computing column-level metadata in this step. As such, instead of passing Dask DFs to the postprocess functions, we can run one top-level `map_batches` call that applies the transformations to batches as pandas dataframes.
> 
> Concretely:
> 
>     1. Change the `postprocess` function in `postprocessing.py` to something like this:
> 
> 
> ```
> def posproc_preds(df):
>         for of_name, output_feature in output_features.items():
>             return output_feature.postprocess_predictions(
>                 df,
>                 training_set_metadata[of_name],
>             )
>     
>     predictions = backend.df_engine.map_batches(predictions, posproc_preds)
> ```
> 
>     2. Change every `postprocess_predictions` function by removing `output_directory` and `backend` from the signature.
> 
>     3. Change all the calls to `backend.df_engine.map_objects` in `postprocess_predictions` to simply call `df.map`, since it's just a Pandas DF.

I really like this suggestion! I'll try it out now. Thanks travis

File: ludwig/data/postprocessing.py
Comment (tgaddair) [inline ludwig/data/postprocessing.py:53]: Remove `output_directory` and `backend`

File: ludwig/data/dataframe/base.py
Comment (tgaddair) [inline ludwig/data/dataframe/base.py:50]: First element should be a dataframe, not a series.

**Review by tgaddair — APPROVED** [review]
LGTM! Please also double-check performance on higgs with this new approach before landing.

**Comment by magdyksaleh** [comment]
This is ready to merge pending CI passing

**Pull request merged**
